File: Assets/scenes/cometScene/scripts/RailsShipControl.js
#pragma strict

var stickPos : Vector3;
var angleClampMin : Vector3;
var angleClampMax : Vector3;
var xVelocityLimit : float = 10;
var yVelocityLimit : float = 10;

var moveScale : float = 0.7f;

private var rotX : float;
private var rotY : float;
private var rotZ : float;

private var theShip : Transform;
private var initialShipRotation : Vector3;

var stability : float = 1.0f;
var speed : float = 1.0f;


/* lock the ship to this object, halt its movement*/
function Start () {
	theShip = GameObject.Find("TheShip").transform;
	theShip.parent = transform;
	theShip.rigidbody.constraints = RigidbodyConstraints.FreezePosition;
	initialShipRotation = theShip.localEulerAngles;
	PropulsionSystem.instance.enableSystem();
}

function OnDestroy(){
	theShip.parent = null;
	theShip.rigidbody.constraints = RigidbodyConstraints.None;

}

function FixedUpdate () {
	newMove();

}

function newMove(){
	stickPos = PropulsionSystem.instance.joyPos;
	//read the current rotation of the ship and project it onto our xy plane
	var shipRot : Vector3 = theShip.TransformDirection(Vector3.forward);
	var xVal = Vector3.Project(shipRot, transform.right);
	var yVal = Vector3.Project(shipRot, transform.up);
	
	transform.position += new Vector3(xVal.x, yVal.y, 0) * moveScale;
	
	var predictedFwd : Vector3 = Quaternion.AngleAxis(
         theShip.rigidbody.angularVelocity.magnitude * Mathf.Rad2Deg * stability / speed,
         theShip.rigidbody.angularVelocity
     ) * -theShip.transform.forward;
 
     var torqueVector : Vector3 = Vector3.Cross(predictedFwd, Vector3.forward);
     theShip.rigidbody.AddTorque(torqueVector * speed * speed);
     
	
	
	
}

function oldMove(){
	stickPos = PropulsionSystem.instance.joyPos;
	rotX += stickPos.x * 1.0f;
	rotY += stickPos.y * 1.0f;
	rotZ += stickPos.z;
	
	rotX = Mathf.LerpAngle(rotX, 0, 0.01f);	
	rotY = Mathf.LerpAngle(rotY, 0, 0.01f);
	
	
	
    rotY = Mathf.Clamp (rotY, angleClampMin.y, angleClampMax.y);
	rotX = Mathf.Clamp (rotX, angleClampMin.x, angleClampMax.x);
	var newRot : Vector3 = new Vector3(rotY, rotX, rotZ);
	            
    theShip.localEulerAngles =  initialShipRotation + newRot;
    
    var newPos : Vector3 = Vector3(newRot.y, -newRot.x, 0) * 0.01f;
    newPos = theShip.TransformDirection(newPos);
    newPos.z = 0;
    newPos.x = Mathf.Clamp(newPos.x, -xVelocityLimit, xVelocityLimit);
    newPos.y = Mathf.Clamp(newPos.y, -yVelocityLimit, yVelocityLimit);
    transform.position += newPos;
}

File: Assets/scenes/cometScene/scripts/RailsShipControl.cs
using UnityEngine;
using System;


public class RailsShipControl:MonoBehaviour{
	
	public Vector3 stickPos;	//joystick position from pilots console
	public Vector3 angleClampMin;
	public Vector3 angleClampMax;
	public float xVelocityLimit = 10.0f;
	public float yVelocityLimit = 10.0f;

	public float moveScale = 0.7f;
	//do we restrict movement to a plane?
	public bool restrictMovement = true;
	//do we force the ship to rotate to our forward?
	public bool restrictRotation = true;

	Vector3 bitsOffset;
	Transform bits;

	public Transform movementDirection;
	
	float rotX;
	float rotY;
	float rotZ;
	
	Transform theShip;
	Vector3 initialShipRotation;
	
	public float stability = 1.0f;
	public float speed = 1.0f;
	
	
	/* lock the ship to this object, halt its movement*/
	public void Start() {
		theShip = GameObject.Find("TheShip").transform;
		theShip.parent = transform;
		theShip.rigidbody.constraints = RigidbodyConstraints.FreezePosition;
		initialShipRotation = theShip.localEulerAngles;
		PropulsionSystem.instance.enableSystem();

		if(movementDirection == null){
			GameObject go=  new GameObject();
			go.name = "MovementDirection";
			go.transform.parent = transform;
			movementDirection = go.transform;
		}

		bits = transform.Find ("Bits");
		bitsOffset = bits.localPosition;
	}

	public void setDirection(Quaternion direction){
		movementDirection.rotation = direction;


	}
	
	public void OnDestroy(){
		theShip.parent = null;
		theShip.rigidbody.constraints = RigidbodyConstraints.None;
	
	}
	
	public void FixedUpdate() {
		if(restrictMovement) newMove();
		if(restrictRotation) doTorque();

		bits.position = transform.position + movementDirection.TransformDirection(bitsOffset);
		bits.LookAt(transform.position);
	}

	private void doTorque(){
		Vector3 predictedFwd = Quaternion.AngleAxis(
			theShip.rigidbody.angularVelocity.magnitude * Mathf.Rad2Deg * stability / speed,
			theShip.rigidbody.angularVelocity
			) * -theShip.transform.forward;
		
		Vector3 torqueVector = Vector3.Cross(predictedFwd, movementDirection.forward);
		theShip.rigidbody.AddTorque(torqueVector * speed * speed);
	}
	
	public void newMove(){
		stickPos = PropulsionSystem.instance.joyPos;
		//read the current rotation of the ship and project it onto our xy plane



		Vector3 shipRot = theShip.TransformDirection(Vector3.forward);
		Vector3 xVal = Vector3.Project(shipRot, movementDirection.right);
		Vector3 yVal = Vector3.Project(shipRot, movementDirection.up);
		
		transform.position += new Vector3(xVal.x, yVal.y, 0.0f) * moveScale;


	}


	public void OnDrawGizmos(){
		if(movementDirection != null){
			Gizmos.DrawLine(transform.position, transform.position + movementDirection.TransformDirection (-Vector3.forward*10));
		}
	}

}
